Plugin/Tax/Config/ShippingTaxPlugin.php

```diff
@@ -0,0 +1,216 @@
+     *
+     * @param                                          $subject
+     * @param                                          $result
+     * @param ShippingAssignmentInterface              $shippingAssignment
+     * @param \Magento\Quote\Model\Quote\Address\Total $total
+     * @param                                          $useBaseCurrency
+     *
+     * @return null
+     *
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    public function afterGetShippingDataObject(
+        $subject,
+        $result,
+        ShippingAssignmentInterface $shippingAssignment,
+        \Magento\Quote\Model\Quote\Address\Total $total,
+        $useBaseCurrency
+    ) {
+        if ($result === null) {
+            return $result; // no shipping tax calculation
+        }
+
+        $quoteItems = $shippingAssignment->getItems();
+        if (count($quoteItems) === 0) {
+            return $result;  // no products -> no shipping -> nothing to do
+        }
+
+        $store       = $quoteItems[0]->getQuote()->getStore();
+        $dynamicType = $this->getDynamicShippingConfigPath($store);
+
+        if ($dynamicType === FireGentoSource::DYNAMIC_TYPE_SHIPPING_TAX_DEFAULT) {
+            return $result;  // If the default behaviour was configured -> use default tax class id
+        }
+
+        $taxClassId = false;
+        if ($dynamicType === FireGentoSource::DYNAMIC_TYPE_HIGHEST_PRODUCT_TAX) {
+            $taxClassId = $this->getHighestProductTaxClassId($quoteItems, $store);
+        }
+
+        // If no tax class id was found, use default one
+        if (!$taxClassId) {
+            return $result;
+        }
+
+        return $result->setTaxClassKey(
+            $this->taxClassKeyDataObjectFactory->create()
+                ->setType(TaxClassKeyInterface::TYPE_ID)
+                ->setValue($taxClassId)
+        );
+    }
+
+    /**
+     * Method for getting highest product tax class id
+     *
+     * @param                            $quoteItems
+     * @param null|string|bool|int|Store $store
+     *
+     * @return bool|int
+     *
+     * @throws \Magento\Framework\Exception\LocalizedException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
+     */
+    private function getHighestProductTaxClassId($quoteItems, $store)
+    {
+        $taxClassIds       = [];
+        $highestTaxClassId = false;
+
+        foreach ($quoteItems as $quoteItem) {
+
+            /** @var $quoteItem Item */
+            if ($quoteItem->getParentItem()) {
+                continue;
+            }
+
+            // Retrieve the tax percent
+            $taxPercent = $quoteItem->getTaxPercent();
+            if (!$taxPercent) {
+                $taxPercent = $this->getTaxPercent($quoteItem->getTaxClassId(), $store);
+            }
+
+            // Add the tax class
+            if (($taxPercent) && !in_array($taxPercent, $taxClassIds)) {
```

Comment:
Are you sure the `in_array` check works? Doesn't this need to be `isset`, because the tax percent is the array key?

Reply:
think this is gone